test/integration/master/audit_dynamic_test.go

```diff
@@ -0,0 +1,293 @@
+				return false, nil
+			}
+
+			return true, nil
+		})
+		require.NoError(t, err, finalErr)
+	})
+
+	t.Run("two sink", func(t *testing.T) {
+		_, err := kubeclient.AuditregistrationV1alpha1().AuditSinks().Create(&sink2)
+		require.NoError(t, err, "failed to create audit sink2")
+		t.Log("created audit sink2")
+
+		var finalErr error
+		err = wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+			// reset event list
+			eventList1.Store(auditinternal.EventList{})
+			eventList2.Store(auditinternal.EventList{})
+
+			// perform configmap ops
+			configMapOperations(t, kubeclient)
+
+			// check for corresponding events in both sinks
+			missing, err := checkForEvents(t, eventList1)
+			if err != nil {
+				finalErr = fmt.Errorf("%v: failed to match all expected events for sink1, events %#v not found", err, missing)
+				return false, nil
+			}
+
+			missing, err = checkForEvents(t, eventList2)
+			if err != nil {
+				finalErr = fmt.Errorf("%v: failed to match all expected events for sink1, events %#v not found", err, missing)
+				return false, nil
+			}
+
+			return true, nil
+		})
+		require.NoError(t, err, finalErr)
+	})
+
+	t.Run("delete sink", func(t *testing.T) {
+		err := kubeclient.AuditregistrationV1alpha1().AuditSinks().Delete("test2", &metav1.DeleteOptions{})
+		require.NoError(t, err, "failed to delete audit sink2")
+		t.Log("deleted audit sink2")
+
+		var finalErr error
+		err = wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+			// reset event lists
+			eventList1.Store(auditinternal.EventList{})
+			eventList2.Store(auditinternal.EventList{})
+
+			// perform configmap ops
+			configMapOperations(t, kubeclient)
+
+			// check for corresponding events in sink1
+			missing, err := checkForEvents(t, eventList1)
+			if err != nil {
+				finalErr = fmt.Errorf("%v: failed to match all expected events for sink1, events %#v not found", err, missing)
+				return false, nil
+			}
+
+			// check that sink2 is empty
+			if len(eventList2.Load().(auditinternal.EventList).Items) != 0 {
+				finalErr = fmt.Errorf("eventList2 should be empty")
+				return false, nil
+			}
+			return true, nil
+		})
+		require.NoError(t, err, finalErr)
+	})
+
+	t.Run("update sink", func(t *testing.T) {
+		sink1, err := kubeclient.AuditregistrationV1alpha1().AuditSinks().Get("test1", metav1.GetOptions{})
+		require.NoError(t, err)
+		sink1.Spec.Webhook.ClientConfig.URL = &server2.URL
+		_, err = kubeclient.AuditregistrationV1alpha1().AuditSinks().Update(sink1)
+		require.NoError(t, err, "failed to update audit sink1")
+		t.Log("updated audit sink1")
+
+		var finalErr error
+		err = wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+			// reset event lists
+			eventList1.Store(auditinternal.EventList{})
+			eventList2.Store(auditinternal.EventList{})
+
+			// perform configmap ops
+			configMapOperations(t, kubeclient)
```

Comment:
polls within polls are fun :)

this still seems potentially flaky. an event from the previous outer loop could sneak into the store, and if it did, checkForEvents would fail.

it might make more sense to do these checks in order:
1. is the sink actively receiving events (does the store have a non-zero number of items)
2. is the sink caught up (is the last event in the store the last thing we did)
3. did the sink see everything we expected? (the current checkForEvents helper)

we could do those checks like this after we register or change a sink:
1. create and delete a configmap with an incrementing name until the sink is non-empty
2. poll until the last item in the sink is the response event of the delete call for the last configmap we deleted
3. clear the sink store, do configMapOperations, call checkForEvents and poll until the sink observes the events